### PR TITLE
Combine low selectivity vectors generated by the hash join filter

### PR DESCRIPTION
**Pull request by zhli1142015**
Combine low selectivity vectors generated by join filter
**Problem**
We found in TPCDS query72 that the join filter leads to a large number of 
low-selectivity result vectors, which affects the performance of subsequent 
operations.

Details:
The number of input vectors on the probe side of the corresponding join is 
84,087, with a total of 743,851,486 rows (our batch size is set to 10,240). Due 
to a large number of duplicate rows on the build side, the final result 
inflates. The number of output vectors from the join is 2,806,054. The 
corresponding join filter filters out a large portion of the results, so the
 number of output rows is 1,430,253,235. This leads to the output of many 
 sparse vectors (the average batch row count is 504).

 **Solution**
The original logic continues the [loop](https://github.com/facebookincubator/velox/blob/main/velox/exec/HashProbe.cpp#L1065) to fill more rows if no rows pass the
filters. To resolve this issue, we can extend it to handle cases where only 
partial rows pass the filters. We need to ensure that the indices 
'outputRowMapping_' and 'outputTableRows_' are filled as much as possible until 
we either reach the preferred batch size or have processed all rows in the 
current input vector.

This approach will not only address the issue mentioned above but also avoid 
unnecessary data copying.

### CODE REVIEW

**Comment by zhli1142015** [comment]
In our experiment for TPCDS query72, we observed that due to the reduced number of vectors, there is a noticeable performance improvement in operators after the first join and the query latency improves 45% also(57s --> 31s).

**Before this change**
Join generates low seletivity vectors ( probe vectors: 84087 --> output vectors: 2806054)
image

Op after above Join
image

**With this changer**
probe vectors: 84087 --> output vectors: 324659
image
Op after above Join: the total probe time reduced from 2.5 min to 1.1 min
image

**Comment by zhli1142015** [comment]
@mbasmanova  and @Yuhta , could you help toa take a look?
Thanks.

**Comment by zhli1142015** [comment]
CI failure is caused by https://github.com/facebookincubator/velox/issues/10871, thanks.

File: velox/exec/HashProbe.cpp [hunk 1]
```diff
@@ -1079,6 +1087,30 @@ RowVectorPtr HashProbe::getOutputInternal(bool toSpillOutput) {
       return nullptr;
     }
 
+    if (accumulatedNumOutput > 0 ||
+        (filter_ != nullptr && numOut < numOutputBeforeFilter / 10 &&
+         !emptyBuildSide)) {
+      accumulatedNumOutput += numOut;
+      if (!resultIter_->atEnd() &&
+          accumulatedNumOutput < outputTableRowsCapacity_ &&
```
Comment (Yuhta) [inline velox/exec/HashProbe.cpp:1095]: `s/outputTableRowsCapacity_/outputBatchSize/` since `outputTableRowsCapacity_` counts extra space reserved for carryover row

File: velox/exec/HashProbe.cpp [hunk 2]
```diff
@@ -1079,6 +1087,30 @@ RowVectorPtr HashProbe::getOutputInternal(bool toSpillOutput) {
       return nullptr;
     }
 
+    if (accumulatedNumOutput > 0 ||
+        (filter_ != nullptr && numOut < numOutputBeforeFilter / 10 &&
```
Comment (Yuhta) [inline velox/exec/HashProbe.cpp:1091]: How is `/ 10` decided?  Can we increase it a little bit, say `/ 2`?

File: velox/exec/HashProbe.cpp [hunk 3]
```diff
@@ -1431,21 +1472,23 @@ int32_t HashProbe::evalFilter(int32_t numRows) {
     static const char* kPassed = "passed";
 
     if (nullAware_) {
-      leftSemiProjectIsNull_.resize(numRows);
-      leftSemiProjectIsNull_.clearAll();
+      leftSemiProjectIsNull_.resize(numRows + offset, false);
+      if (offset == 0) {
```
Comment (Yuhta) [inline velox/exec/HashProbe.cpp:1476]: Since left semi does not filter probe rows, we should always get `offset == 0` here, no?  If that's the case just do `VELOX_CHECK_EQ(offset, 0)`; otherwise `leftSemiProjectIsNull_` should be initialized outside the loop.

**Review by Yuhta — COMMENTED** [review]
I think the idea is great, leave some initial comments.

Related issue #7801

File: velox/vector/BaseVector.h [hunk 4]
```diff
@@ -863,6 +863,18 @@ class BaseVector {
     storageByteCount_ = std::nullopt;
   }
 
+  // Slice a buffer with specific type.
+  //
+  // For boolean type and if the offset is not multiple of 8, return a shifted
+  // copy; otherwise return a BufferView into the original buffer (with shared
+  // ownership of original buffer).
+  static BufferPtr sliceBuffer(
```
Comment (mbasmanova) [inline velox/vector/BaseVector.h:871]: Are you making this API public? If so, please, extract this change into a separate PR and add tests.

File: velox/vector/BaseVector.h [hunk 5]
```diff
@@ -863,6 +863,18 @@ class BaseVector {
...
+      const Type&,
```
Comment (mbasmanova) [inline velox/vector/BaseVector.h:872]: please, add argument names and clarify that 'pool' is used to allocate new buffer for bool case.

File: velox/exec/tests/HashJoinTest.cpp [hunk 6]
```diff
@@ -8112,4 +8112,85 @@ TEST_F(HashJoinTest, nanKeys) {
...
+        .hashJoin(
+            {"t0"},
+            {"u0"},
+            PlanBuilder(planNodeIdGenerator)
+                .tableScan(asRowType(buildVectors[0]->type()))
+                .capturePlanNodeId(buildScanId)
+                .planNode(),
+            "(t1 + u1) % 3 = 0",
+            {"t0", "t1", "match"},
+            core::JoinType::kLeftSemiProject,
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:8163]: Why kLeftSemiProject ? Should we tests all join types?

File: velox/exec/tests/HashJoinTest.cpp [hunk 7]
```diff
@@ -8112,4 +8112,85 @@ TEST_F(HashJoinTest, nanKeys) {
...
+
+    HashJoinBuilder(*pool_, duckDbQueryRunner_, driverExecutor_.get())
+        .planNode(plan)
+        .inputSplits(splitInput)
+        .checkSpillStats(false)
+        .referenceQuery(
+            "SELECT t0, t1, t0 IN (SELECT u0 FROM u WHERE (t1 + u1) % 3 = 0) FROM t")
+        .verifier([&](const std::shared_ptr<Task>& task, bool /*unused*/) {
+          auto stats = task->taskStats();
+          for (auto& pipeline : stats.pipelineStats) {
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:8185]: Let's use toPlanStats(stats)

File: velox/exec/tests/HashJoinTest.cpp [hunk 8]
```diff
@@ -8112,4 +8112,85 @@ TEST_F(HashJoinTest, nanKeys) {
        makeFlatVector<int64_t>({1, 2})});
   facebook::velox::test::assertEqualVectors(expected, result);
 }
+
+TEST_F(HashJoinTest, combineSmallVectorsAfterFilter) {
+  // Verify low selectivity / small vectors are combined to 1 vector.
```
Comment (mbasmanova) [inline velox/exec/tests/HashJoinTest.cpp:8117]: Would you add a comment to describe the shape of the data required for this test?

File: velox/vector/BaseVector.h (hunk 4 above)
Comment (zhli1142015) [reply velox/vector/BaseVector.h:871]: Create PR for this: https://github.com/facebookincubator/velox/pull/11001, please help to take a look, thanks.

File: velox/exec/HashProbe.cpp (hunk 3 above)
Comment (zhli1142015) [reply velox/exec/HashProbe.cpp]: I actually observe the case `offset > 0` in UT.
I moved the intialization for `leftSemiProjectIsNull_` outside the loop. Thanks.

File: velox/exec/HashProbe.cpp (hunk 2 above)
Comment (zhli1142015) [reply velox/exec/HashProbe.cpp]: Updated, thanks

File: velox/exec/HashProbe.cpp (hunk 1 above)
Comment (zhli1142015) [reply velox/exec/HashProbe.cpp]: I double checked that `accumulatedNumOutput < outputTableRowsCapacity_` and the condiction `outputBatchSize - numOut > 0` are duplicate. I only kept the second condition.
Thanks.

File: velox/exec/tests/HashJoinTest.cpp (hunk 6 above)
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp:8163]: Sure, will add tests for all join types.

Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp]: Updated, thanks.

File: velox/exec/tests/HashJoinTest.cpp (hunk 7 above)
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp]: Updated, thanks.

File: velox/exec/tests/HashJoinTest.cpp (hunk 8 above)
Comment (zhli1142015) [reply velox/exec/tests/HashJoinTest.cpp]: Updated, thanks.

File: velox/exec/HashProbe.cpp (hunk 3 above)
Comment (Yuhta) [reply velox/exec/HashProbe.cpp]: Curious when does this happen?

Comment (zhli1142015) [reply velox/exec/HashProbe.cpp]: The test case I added for semi project join can repro it.
`SELECT t0, t1, t0 IN (SELECT u0 FROM u WHERE (t1 + u1) % 3 = 0) FROM t`
I think we also need to get the results by function `listJoinResults` and `evalFilter` for left semi join with filter, when we merge small vectors, we need to set correct offset for `leftSemiProjectIsNull_`.

File: velox/exec/HashProbe.cpp [hunk 9]
```diff
@@ -1079,6 +1094,34 @@ RowVectorPtr HashProbe::getOutputInternal(bool toSpillOutput) {
...
+
+      // To avoid generating low seletivity / small vectors, continue the
+      // current loop to populate more results to 'outputRowMapping_' and
+      // 'outputTableRows_' until all rows in the current input are processed or
+      // the preferred number of rows has been produced.
+      if (!resultIter_->atEnd() &&
+          accumulatedNumOutput < operatorCtx_->driverCtx()
+                                     ->queryConfig()
+                                     .preferredOutputBatchBytes() &&
+          outputBatchSize - numOut > 0) {
```
Comment (Yuhta) [inline velox/exec/HashProbe.cpp:1110]: nit: `outputBatchSize > numOut`

File: velox/exec/HashProbe.cpp (hunk 1 above)
Comment (Yuhta) [reply velox/exec/HashProbe.cpp]: Can you explain why those 2 are duplicate? With
* numOut <= accumulatedNumOutput
* outputBatchSize <= outputTableRowsCapacity_
* numOut < outputBatchSize

how do you deduct
* accumulatedNumOutput < outputTableRowsCapacity_

?

File: velox/exec/HashProbe.cpp (hunk 3 above)
Comment (Yuhta) [reply velox/exec/HashProbe.cpp]: I see, there is still a filter, thanks

File: velox/exec/HashProbe.cpp (hunk 1 above)
Comment (zhli1142015) [reply velox/exec/HashProbe.cpp]: Before entering the for loop on line 1023, `outputTableRowsCapacity_` is set to `outputBatchSize + 1`. In each iteration, `numOut` is added to `accumulatedNumOutput`, and `numOut` is subtracted from `outputBatchSize`. Both conditions are  for checking if `outputTableRows` has sufficient capacity to ontinue. I think these two Therefore, I believe only the condition `outputBatchSize - numOut > 0` should be kept.

**Comment by zhli1142015** [comment]
Hello @Yuhta and @mbasmanova , could you take a look again? Thanks.

File: velox/vector/BaseVector.h [hunk 10]
```diff
@@ -867,6 +867,18 @@ class BaseVector {
     storageByteCount_ = std::nullopt;
   }
 
+  // Slice a buffer with specific type.
+  //
+  // For boolean type and if the offset is not multiple of 8, return a shifted
+  // copy; otherwise return a BufferView into the original buffer (with shared
+  // ownership of original buffer).
+  static BufferPtr sliceBuffer(
```
Comment (Yuhta) [inline velox/vector/BaseVector.h:875]: We can remove this

File: velox/exec/HashProbe.cpp [hunk 11]
```diff
@@ -998,19 +998,28 @@ RowVectorPtr HashProbe::getOutputInternal(bool toSpillOutput) {
...
+        isAntiJoin(joinType_)) {
+      // If we need non-matching probe side row, there is a possibility that
+      // such row exists at end of an input batch and being carried over in the
+      // next output batch, so we need to make extra room of one row in output.
+      ++outputTableRowsCapacity_;
+    }
+
+    // Intialize 'leftSemiProjectIsNull_' for null aware lft semi join.
+    if (isLeftSemiProjectJoin(joinType_) && nullAware_) {
+      leftSemiProjectIsNull_.clearAll();
```
Comment (Yuhta) [inline velox/exec/HashProbe.cpp:1012]: Can we do `leftSemiProjectIsNull_.resize(outputTableRowsCapacity_)` here so we don't need to do it in each loop?  `SelectivityVector::resize` requires iterate over all the bits in it.

**Comment by zhli1142015** [comment]
CI failure is not related to this PR, thanks.

**Comment by zhli1142015** [comment]
Kindly ping, @Yuhta , thanks.

File: velox/vector/BaseVector.h (hunk 10 above)
Comment (zhli1142015) [reply velox/vector/BaseVector.h]: Thanks, removed.

**Comment by zhli1142015** [comment]
Hello @Yuhta  and @mbasmanova ,
Do you have more comments?
Thanks.

File: velox/exec/HashProbe.cpp (hunk 11 above)
Comment (Yuhta) [reply velox/exec/HashProbe.cpp:1021]: Looks like we are still resizing it in `evalFilter`?

File: velox/exec/HashProbe.cpp [hunk 12]
```diff
@@ -1440,21 +1488,20 @@ int32_t HashProbe::evalFilter(int32_t numRows) {
     static const char* kPassed = "passed";
 
     if (nullAware_) {
-      leftSemiProjectIsNull_.resize(numRows);
-      leftSemiProjectIsNull_.clearAll();
+      leftSemiProjectIsNull_.resize(numRows + offset, false);
```
Comment (Yuhta) [inline velox/exec/HashProbe.cpp:1491]: We can initialize it once outside the loop. Each loop will only write into non-overlapping subranges here.

Comment (zhli1142015) [reply velox/exec/HashProbe.cpp]: I see, updated.

File: velox/exec/HashProbe.cpp (hunk 11 above)
Comment (zhli1142015) [reply velox/exec/HashProbe.cpp:1022]: Moved it outside the loop, thanks.

**Comment by zhli1142015** [comment]
Failed test should not be related to this PR.

```
[ RUN      ] TaskPauseTest.resumeFuture
/home/runner/work/velox/velox/velox/velox/exec/tests/TaskTest.cpp:1545: Failure
Expected equality of these values:
  task_->numFinishedDrivers()
    Which is: 0
  1
[  FAILED  ] TaskPauseTest.resumeFuture (108 ms)

[ RUN      ] AsyncDataCacheTest/AsyncDataCacheTest.ttl/0
I20241114 08:55:17.632177 21144 SsdCache.cpp:45] [SSDCA] SSD cache config: 4 shards, capacity 128.00MB, checkpoint size 6.40MB, file cow ENABLED, checksum DISABLED, read verification DISABLED
W20241114 08:55:17.632278 21144 SsdFile.cpp:878] [SSDCA] Starting shard 0 without checkpoint, with checksum write disabled, read verification disabled, checkpoint file /tmp/velox_test_k2IJrO/cache0.cpt
W20241114 08:55:17.632366 21144 SsdFile.cpp:878] [SSDCA] Starting shard 1 without checkpoint, with checksum write disabled, read verification disabled, checkpoint file /tmp/velox_test_k2IJrO/cache1.cpt
W20241114 08:55:17.632467 21144 SsdFile.cpp:878] [SSDCA] Starting shard 2 without checkpoint, with checksum write disabled, read verification disabled, checkpoint file /tmp/velox_test_k2IJrO/cache2.cpt
W20241114 08:55:17.632550 21144 SsdFile.cpp:878] [SSDCA] Starting shard 3 without checkpoint, with checksum write disabled, read verification disabled, checkpoint file /tmp/velox_test_k2IJrO/cache3.cpt
W20241114 08:55:17.632655 21144 MemoryArbitrator.cpp:84] Query memory capacity[32.00MB] is set for NOOP arbitrator which has no capacity enforcement
I20241114 08:55:17.758888 43268 AsyncDataCache.cpp:573] [SSDCA] Limiting SSD save batch to 280 entries
I20241114 08:55:17.758948 43268 AsyncDataCache.cpp:573] [SSDCA] Limiting SSD save batch to 295 entries
I20241114 08:55:17.758965 43268 AsyncDataCache.cpp:573] [SSDCA] Limiting SSD save batch to 277 entries
I20241114 08:55:17.758972 43268 AsyncDataCache.cpp:573] [SSDCA] Limiting SSD save batch to 282 entries
I20241114 08:55:17.762882 43315 SsdFile.cpp:345] [SSDCA] Grow cache file /tmp/velox_test_k2IJrO/cache0 to 1 regions (max: 1)
I20241114 08:55:17.764694 43317 SsdFile.cpp:345] [SSDCA] Grow cache file /tmp/velox_test_k2IJrO/cache1 to 1 regions (max: 1)
I20241114 08:55:17.769876 43318 SsdFile.cpp:345] [SSDCA] Grow cache file /tmp/velox_test_k2IJrO/cache2 to 1 regions (max: 1)
I20241114 08:55:17.775424 43319 SsdFile.cpp:345] [SSDCA] Grow cache file /tmp/velox_test_k2IJrO/cache3 to 1 regions (max: 1)
I20241114 08:55:17.779536 43319 SsdCache.cpp:138] [SSDCA] Wrote 2.97MB, inf bytes/s
I20241114 08:55:17.878214 43256 AsyncDataCache.cpp:573] [SSDCA] Limiting SSD save batch to 358 entries
I20241114 08:55:17.878252 43256 AsyncDataCache.cpp:573] [SSDCA] Limiting SSD save batch to 366 entries
I20241114 08:55:17.878265 43256 AsyncDataCache.cpp:573] [SSDCA] Limiting SSD save batch to 370 entries
I20241114 08:55:17.895213 43372 SsdFile.cpp:747] [SSDCA] Checkpointing shard 1, force: 0 bytesAfterCheckpoint: 1.95MB checkpointIntervalBytes: 1.60MB
I20241114 08:55:17.898605 43374 SsdFile.cpp:747] [SSDCA] Checkpointing shard 2, force: 0 bytesAfterCheckpoint: 1.63MB checkpointIntervalBytes: 1.60MB
I20241114 08:55:17.901418 43374 SsdFile.cpp:857] [SSDCA] Checkpoint persisted with 163 cache entries
I20241114 08:55:17.905488 43377 SsdFile.cpp:747] [SSDCA] Checkpointing shard 3, force: 0 bytesAfterCheckpoint: 1.68MB checkpointIntervalBytes: 1.60MB
I20241114 08:55:17.911123 43377 SsdFile.cpp:857] [SSDCA] Checkpoint persisted with 168 cache entries
I20241114 08:55:17.919091 43372 SsdFile.cpp:857] [SSDCA] Checkpoint persisted with 195 cache entries
I20241114 08:55:17.919142 43372 SsdCache.cpp:138] [SSDCA] Wrote 3.70MB, inf bytes/s
I20241114 08:55:17.927124 43385 AsyncDataCache.cpp:618] [CACHE] Removed 0 AsyncDataCache entries.
I20241114 08:55:17.929157 43385 AsyncDataCache.cpp:618] [CACHE] Removed 510 AsyncDataCache entries.
I20241114 08:55:17.932183 43385 AsyncDataCache.cpp:618] [CACHE] Removed 517 AsyncDataCache entries.
I20241114 08:55:17.935236 43385 AsyncDataCache.cpp:618] [CACHE] Removed 522 AsyncDataCache entries.
I20241114 08:55:17.936296 43385 SsdFile.cpp:671] [SSDCA] Removed 141 entries from /tmp/velox_test_k2IJrO/cache0. And erased 1 regions with 50% entries removed, and 0 left.
I20241114 08:55:17.936650 43385 SsdFile.cpp:671] [SSDCA] Removed 195 entries from /tmp/velox_test_k2IJrO/cache1. And erased 1 regions with 50% entries removed, and 0 left.
I20241114 08:55:17.936947 43385 SsdFile.cpp:671] [SSDCA] Removed 163 entries from /tmp/velox_test_k2IJrO/cache2. And erased 1 regions with 50% entries removed, and 0 left.
I20241114 08:55:17.937261 43385 SsdFile.cpp:671] [SSDCA] Removed 168 entries from /tmp/velox_test_k2IJrO/cache3. And erased 1 regions with 50% entries removed, and 0 left.
I20241114 08:55:17.937273 43385 CacheTTLController.cpp:68] Succeeded applying cache TTL of 98 seconds. Entries from 66 files are to be removed, while 0 files are retained
I20241114 08:55:17.937394 43384 AsyncDataCache.cpp:618] [CACHE] Removed 497 AsyncDataCache entries.
I20241114 08:55:17.938243 43384 AsyncDataCache.cpp:618] [CACHE] Removed 0 AsyncDataCache entries.
I20241114 08:55:17.938344 43384 AsyncDataCache.cpp:618] [CACHE] Removed 0 AsyncDataCache entries.
I20241114 08:55:17.938447 43384 AsyncDataCache.cpp:618] [CACHE] Removed 0 AsyncDataCache entries.
I20241114 08:55:17.938455 43384 SsdFile.cpp:671] [SSDCA] Removed 0 entries from /tmp/velox_test_k2IJrO/cache0. And erased 0 regions with 50% entries removed, and 0 left.
I20241114 08:55:17.938464 43384 SsdFile.cpp:671] [SSDCA] Removed 0 entries from /tmp/velox_test_k2IJrO/cache1. And erased 0 regions with 50% entries removed, and 0 left.
I20241114 08:55:17.938472 43384 SsdFile.cpp:671] [SSDCA] Removed 0 entries from /tmp/velox_test_k2IJrO/cache2. And erased 0 regions with 50% entries removed, and 0 left.
I20241114 08:55:17.938480 43384 SsdFile.cpp:671] [SSDCA] Removed 0 entries from /tmp/velox_test_k2IJrO/cache3. And erased 0 regions with 50% entries removed, and 0 left.
I20241114 08:55:17.938488 43384 CacheTTLController.cpp:68] Succeeded applying cache TTL of 98 seconds. Entries from 66 files are to be removed, while 0 files are retained
/home/runner/work/velox/velox/velox/velox/common/caching/tests/AsyncDataCacheTest.cpp:1354: Failure
Expected equality of these values:
  statsTtl.ssdStats->entriesAgedOut
    Which is: 667
  statsT1.ssdStats->entriesCached
    Which is: 297
I20241114 08:55:17.938970 21144 SsdCache.cpp:209] [SSDCA] SSD cache is shutting down
I20241114 08:55:17.938992 21144 SsdFile.cpp:747] [SSDCA] Checkpointing shard 0, force: 1 bytesAfterCheckpoint: 1.41MB checkpointIntervalBytes: 1.60MB
I20241114 08:55:17.941179 21144 SsdFile.cpp:857] [SSDCA] Checkpoint persisted with 0 cache entries
I20241114 08:55:17.941211 21144 SsdFile.cpp:747] [SSDCA] Checkpointing shard 1, force: 1 bytesAfterCheckpoint: 0B checkpointIntervalBytes: 1.60MB
I20241114 08:55:17.941892 21144 SsdFile.cpp:857] [SSDCA] Checkpoint persisted with 0 cache entries
I20241114 08:55:17.941926 21144 SsdFile.cpp:747] [SSDCA] Checkpointing shard 2, force: 1 bytesAfterCheckpoint: 0B checkpointIntervalBytes: 1.60MB
I20241114 08:55:17.942683 21144 SsdFile.cpp:857] [SSDCA] Checkpoint persisted with 0 cache entries
I20241114 08:55:17.942718 21144 SsdFile.cpp:747] [SSDCA] Checkpointing shard 3, force: 1 bytesAfterCheckpoint: 0B checkpointIntervalBytes: 1.60MB
I20241114 08:55:17.943461 21144 SsdFile.cpp:857] [SSDCA] Checkpoint persisted with 0 cache entries
I20241114 08:55:17.943491 21144 SsdCache.cpp:216] [SSDCA] SSD cache has been shutdown
I20241114 08:55:17.953382 21144 TempDirectoryPath.cpp:29] TempDirectoryPath:: removing all files from /tmp/velox_test_k2IJrO
[  FAILED  ] AsyncDataCacheTest/AsyncDataCacheTest.ttl/0, where GetParam() = 2-byte object <00-00> (321 ms)
```

**Review by Yuhta — APPROVED** [review]
@zhli1142015 Can you create 2 issues with the failing test names in it (if they do not exist yet)?
https://github.com/facebookincubator/velox/actions/runs/11833246203/job/32971531602

**Comment by zhli1142015** [comment]
Created issue: https://github.com/facebookincubator/velox/issues/11550 for the CI failures, thanks.

**Comment by tanjialiang** [comment]
copied from https://github.com/facebookincubator/velox/pull/11659

We observed memory leak crashes in Meta's clusters due to this PR. We confirmed that PR was the offending PR by comparing binary w/wo that PR on a repro query.
I have commented in the revert PR (#11659) of one bug I have identified. But I am not sure if this is the bug causing leak since I have not investigated further.
